### ISSUE TITLE
shouldEstablishDirectChannel not establishing connection on first app launch

### ISSUE THREAD

**Issue by alanscarpa** [firebase/firebase-ios-sdk#210]
### Describe your environment

  * Xcode version: 8.3.3
  * Firebase 4.1.0  
  * FirebaseAnalytics 4.0.3  
  * FirebaseCore 4.0.5 
  * FirebaseInstanceID 2.0.1 
  * FirebaseMessaging 2.0.1
  * Firebase Product: Messaging 

###  Describe the problem
When launching an app for the first time, or after deleting and re-installing, `Messaging.messaging().shouldEstablishDirectChannel` does not establish the socket connection.  If I shut down the app, and re-open it, then the socket connection is established.

#### Steps to reproduce:
1)  Place this code in AppDelegate:
```
FirebaseApp.configure()
Messaging.messaging().delegate = self
Messaging.messaging().shouldEstablishDirectChannel = true
````
2)  Place this code anywhere after to check if connection is established:
`Messaging.messaging().isDirectChannelEstablished` 
This always returns false.
3) Listen for connection state change and observe that this notification never gets fired.
```
NotificationCenter.default.addObserver(self, selector:
            #selector(fcmConnectionStateChange), name:
            NSNotification.Name.MessagingConnectionStateChanged, object: nil)
```

That is the problem in a nutshell.  If I simply kill the app, and re-launch it, everything works as expected.  The socket connection is made and the `MessagingConnectionStateChanged` notification is fired.

Why is `Messaging.messaging().shouldEstablishDirectChannel ` not connecting on my initial app launch?

#### Relevant Code:

```
func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        window = UIWindow(frame: UIScreen.main.bounds)
        window!.rootViewController = RootViewController.shared
        window!.makeKeyAndVisible()

        setUpFirebase()
        setUpPushNotificationsForApplication(application)
        RootViewController.shared.goToLoginVC()
        
        return true
    }
    
    // MARK: - Firebase
    
    func setUpFirebase() {
        NotificationCenter.default.addObserver(self, selector:
            #selector(fcmConnectionStateChange), name:
            NSNotification.Name.MessagingConnectionStateChanged, object: nil)        
        FirebaseApp.configure()
        Messaging.messaging().delegate = self
        Messaging.messaging().shouldEstablishDirectChannel = true
    }
    
    // MARK: - Firebase Notifications
    
    func fcmConnectionStateChange() {
       // This is never called on app's first launch!!!
        print(Messaging.messaging().isDirectChannelEstablished)
    }

```

**Comment by PaulRBerg** [comment]
+1

**Comment by alanscarpa** [comment]
Anyone have a hunch as to why this might be occurring? Are there any steps that should definitely be taken before setting `shouldEstablishDirectChannel = true`?

**Comment by alanscarpa** [comment]
@rsattar Any word on this or if you guys will have a chance to take a look at this?

**Comment by rsattar** [comment]
Hmm, what you're doing makes sense, and that's what I'm doing too, but I'm connection on first launch also:
```
FirebaseApp.configure()
Messaging.messaging().delegate = self
Messaging.messaging().shouldEstablishDirectChannel = true
listenForDirectChannelStateChanges();
```
Where I've got:
```
extension AppDelegate {
  func listenForDirectChannelStateChanges() {
    NotificationCenter.default.addObserver(self, selector: #selector(onMessagingDirectChannelStateChanged(_:)), name: .MessagingConnectionStateChanged, object: nil)
  }

  func onMessagingDirectChannelStateChanged(_ notification: Notification) {
    print("Is FCM Direct Channel Established: \(Messaging.messaging().isDirectChannelEstablished)")
  }
}
```
And I'm being called on the first launch:
```
Is FCM Direct Channel Established: true
```
I'm also breakpointing in `FIRMessagingClient`'s `connect` method, and it is getting called as well.

As a quick check, can you try explicitly setting your `FirebaseInstanceID` pod to point to `2.0.0`? There is a startup regression in `2.0.1` that could be related to first launch experiences.

**Comment by rsattar** [comment]
Okay, I've managed to reproduce the issue, and it _is_ related to the bug in FirebaseInstanceID 2.0.2. A fix for this has already been made and should be getting released in the next release of Firebase.

For now, a quick workaround is to simply add:
```
application.registerForRemoteNotifications()
```
to your App Delegate.

**NOTE**: This is safe to call on app startup, as it does _not_ show the user notifications permissions dialog to the user. The dialog is only shown when you call `UNUserNotificationCenter.current().requestAuthorization(...)` or `application.registerUserNotificationSettings(...)`.

**Comment by alanscarpa** [comment]
Setting `FirebaseInstanceID` to `2.0` fixed the issue.  Thank you!  I'll use this for now.

`application.registerForRemoteNotifications()` was already being called in one of my functions within app delegate (I just think I didn't include in my code snippet).  I moved it to the very top of AppDelegate, still did not work.

What is the best way to stay up to date on Firebase releases so I can update when this issue is fixed in next release?  Thanks for the help @rsattar !

**Comment by rsattar** [comment]
Hmm, that is strange that it didn't fix the issue when you added `registerForRemoteNotifications()` at the top. That should have definitely fixed it, if you were able to get the APNs token.

We have a [release notes page](https://firebase.google.com/support/release-notes/ios) that shows the current releases and all the fixes/updates in them. Hope that helps!

**Comment by alanscarpa** [comment]
It actually worked when I "Allow" pushes.  But if I deny pushes - I do not get the FCM connection on first launch.  Though, in 2.0.0, if I deny pushes, my FCM connection does work on initial launch - which is what I was looking for.

**Comment by rsattar** [comment]
Hmm, thanks for the extra info @alanscarpa. I think I've found another issue which might be causing this!

**Comment by rsattar** [comment]
FYI, I've added a bug fix in #290. Unfortunately it's half of the fix, and the other half is in the Instance ID SDK. These should be available in the release after next, as I missed the window for the next launch.

**Comment by paulb777** [comment]
Fix released today in Firebase 4.3.0, FirebaseMessaging 2.0.4, and FirebaseInstanceID 2.0.4
